### PR TITLE
feat: Support Spark Expression Encode

### LINKED ISSUE #3183: [Feature] Support Spark expression: encode
## What is the problem the feature request solves?

> **Note:** This issue was generated with AI assistance. The specification details have been extracted from Spark documentation and may need verification.

Comet does not currently support the Spark `encode` function, causing queries using this function to fall back to Spark's JVM execution instead of running natively on DataFusion.

The Encode expression converts a string to binary data using a specified character encoding. It is a runtime-replaceable expression that delegates to a static method for actual encoding operations, supporting legacy charset and error handling configurations.

Supporting this expression would allow more Spark workloads to benefit from Comet's native acceleration.

### Spark Specification

**Syntax:**
```sql
ENCODE(string_expr, charset_expr)
```

**Arguments:**
| Argument | Type | Description |
|----------|------|-------------|
| str | String | The input string expression to be encoded |
| charset | String | The character set/encoding name to use for conversion |
| legacyCharsets | Boolean | Internal flag for legacy Java charset handling behavior |
| legacyErrorAction | Boolean | Internal flag for legacy coding error action behavior |

**Return Type:** BinaryType - Returns binary data representing the encoded string.

**Supported Data Types:**
- Input string: StringTypeWithCollation (supports trim collation)
- Input charset: StringTypeWithCollation (supports trim collation)
- Both string inputs support collation-aware string types

**Edge Cases:**
- Null input string or charset results in null output
- Invalid charset names may throw runtime exceptions
- Legacy flags affect error handling behavior for malformed input
- Empty string input produces empty binary output
- Behavior varies based on SQLConf legacy settings for charset and error handling

**Examples:**
```sql
-- Encode string using UTF-8
SELECT ENCODE('hello world', 'UTF-8');

-- Encode with different charset
SELECT ENCODE('café', 'ISO-8859-1');
```

```scala
// DataFrame API usage
import org.apache.spark.sql.functions._
df.select(expr("ENCODE(name, 'UTF-8')").as("encoded_name"))

// Using column expressions
df.select(encode(col("text_column"), lit("UTF-16")))
```

### Implementation Approach

See the [Comet guide on adding new expressions](https://datafusion.apache.org/comet/contributor-guide/adding_a_new_expression.html) for detailed instructions.

1. **Scala Serde**: Add expression handler in `spark/src/main/scala/org/apache/comet/serde/`
2. **Register**: Add to appropriate map in `QueryPlanSerde.scala`
3. **Protobuf**: Add message type in `native/proto/src/proto/expr.proto` if needed
4. **Rust**: Implement in `native/spark-expr/src/` (check if DataFusion has built-in support first)

## Additional context

**Difficulty:** Medium
**Spark Expression Class:** `org.apache.spark.sql.catalyst.expressions.Encode`

**Related:**
- Decode - inverse operation to convert binary back to string
- Cast expressions for type conversions
- String manipulation functions in string_funcs group

---
*This issue was auto-generated from Spark reference documentation.*

### PR DESCRIPTION
**Pull request by YutaLin**
## Which issue does this PR close?

Closes #3183

## Rationale for this change
Support expression Encode

## What changes are included in this PR?
* Add StringEncode in string serde
* Update shims in spark3.4/3.5/4.0/4.1/4.2 to catch Encode

## How are these changes tested?

Add encode.sql and run it in spark 3.4/3.5/4.0

### CODE REVIEW

File: spark/src/main/spark-4.0/org/apache/comet/shims/CometExprShim.scala [hunk 1]
```diff
@@ -93,6 +93,19 @@ trait CometExprShim extends CommonStringExprs {
...
+            s.dataType.isInstanceOf[BinaryType] &&
+            s.functionName == "encode" &&
+            s.arguments.size == 4 &&
+            s.inputTypes == Seq(
+              StringTypeWithCollation(supportsTrimCollation = true),
+              StringTypeWithCollation(supportsTrimCollation = true),
+              BooleanType,
+              BooleanType) =>
+        val Seq(value, charset, _, _) = s.arguments
+        stringEncode(expr, charset, value, inputs, binding)
```
Comment (andygrove) [inline spark/src/main/spark-4.0/org/apache/comet/shims/CometExprShim.scala]: This looks identical for all Spark 4.x versions - could this move to the spark-4.x shim to avoid duplicating?

File: spark/src/main/scala/org/apache/comet/serde/strings.scala [hunk 2]
```diff
@@ -495,4 +495,33 @@ trait CommonStringExprs {
...
+
+  def stringEncode(
+      expr: Expression,
+      charset: Expression,
+      value: Expression,
+      inputs: Seq[Attribute],
+      binding: Boolean): Option[Expr] = {
+    charset match {
+      case Literal(str, DataTypes.StringType)
+          if str.toString.toLowerCase(Locale.ROOT) == "utf-8" =>
```
Comment (andygrove) [inline spark/src/main/scala/org/apache/comet/serde/strings.scala]: Is it possible for `str` to be null? If so, it could cause NPE here

Comment (andygrove) [reply spark/src/main/scala/org/apache/comet/serde/strings.scala]: My AI reviewer said "Spark accepts utf8 as an alias for UTF-8". Could you check if this is accurate? We may need to handle that case as well.

**Comment by andygrove** [comment]
Thanks @YutaLin. LGTM overall. Could you address feedback, then I'll kick off CI

**Comment by YutaLin** [comment]
Hi @andygrove, thanks for the review!
I've extract encode method and add null check.

About "Spark accepts utf8 as an alias for UTF-8", spark only supports alias before 3.5, because it uses JDK `Charset.forName`. After 4.0, it has a whitelist check, so it doesn't support alias. I'd suggest we keep only utf-8 now, WDYT?

https://spark.apache.org/docs/4.0.0/sql-migration-guide.html#upgrading-from-spark-sql-35-to-40
> Since Spark 4.0, the encode() and decode() functions support only the following charsets ‘US-ASCII’, ‘ISO-8859-1’, ‘UTF-8’, ‘UTF-16BE’, ‘UTF-16LE’, ‘UTF-16’, ‘UTF-32’. To restore the previous behavior when the function accepts charsets of the current JDK used by Spark, set spark.sql.legacy.javaCharsets to true.

File: spark/src/main/scala/org/apache/comet/serde/strings.scala [hunk 3]
```diff
@@ -495,4 +495,33 @@ trait CommonStringExprs {
...
+
+  def stringEncode(
+      expr: Expression,
+      charset: Expression,
+      value: Expression,
+      inputs: Seq[Attribute],
+      binding: Boolean): Option[Expr] = {
+    charset match {
+      case Literal(str, DataTypes.StringType)
+          if str != null && str.toString.toLowerCase(Locale.ROOT) == "utf-8" =>
```
Comment (coderfender) [inline spark/src/main/scala/org/apache/comet/serde/strings.scala:670]: This seems like a fragile check . Perhaps a better check would be to use `StandardCharsets` ?

File: spark/src/main/spark-4.x/org/apache/comet/shims/ShimCometExprs.scala [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.expressions.objects.StaticInvoke
+import org.apache.spark.sql.internal.types.StringTypeWithCollation
+import org.apache.spark.sql.types.{BinaryType, BooleanType}
+
+import org.apache.comet.serde.CommonStringExprs
+import org.apache.comet.serde.ExprOuterClass.Expr
+
+trait ShimCometExprs extends CommonStringExprs {
```
Comment (coderfender) [inline spark/src/main/spark-4.x/org/apache/comet/shims/ShimCometExprs.scala]: seems like the trait name seems to be different from `Comet<>` ?

File: spark/src/test/resources/sql-tests/expressions/string/encode.sql [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+--   StaticInvoke(classOf[Encode], BinaryType, "encode", ...)
+
+statement
+CREATE TABLE test_encode_utf8(s string) USING parquet
+
+statement
+INSERT INTO test_encode_utf8 VALUES ('hello'), ('world'), (''), ('café'), (NULL)
+
+query
+SELECT encode(s, 'utf-8') FROM test_encode_utf8
```
Comment (coderfender) [inline spark/src/test/resources/sql-tests/expressions/string/encode.sql:31]: I believe spark accepts any variant of `utf-8` (`utf8 /UTF8` etc) . May be a good idea to add those tests ?

File: spark/src/main/spark-4.1/org/apache/comet/shims/CometExprShim.scala [hunk 6]
```diff
@@ -92,6 +92,19 @@ trait CometExprShim extends CommonStringExprs {
...
+            s.functionName == "encode" &&
+            s.arguments.size == 4 &&
+            s.inputTypes == Seq(
+              StringTypeWithCollation(supportsTrimCollation = true),
+              StringTypeWithCollation(supportsTrimCollation = true),
+              BooleanType,
+              BooleanType) =>
+        val Seq(value, charset, _, _) = s.arguments
+        stringEncode(expr, charset, value, inputs, binding)
+
```
Comment (coderfender) [inline spark/src/main/spark-4.1/org/apache/comet/shims/CometExprShim.scala:68]: Seems like we missed removing this piece ?

**Review by coderfender — COMMENTED** [review]
Left some minor comments but overall looks good @YutaLin

File: spark/src/test/resources/sql-tests/expressions/string/encode.sql (hunk 5 above)
Comment (YutaLin) [reply spark/src/test/resources/sql-tests/expressions/string/encode.sql:31]: uft8/UTF8 only supports version before v4.0
https://github.com/apache/datafusion-comet/pull/4315#issuecomment-4445627791

File: spark/src/main/scala/org/apache/comet/serde/strings.scala (hunk 3 above)
Comment (YutaLin) [reply spark/src/main/scala/org/apache/comet/serde/strings.scala:670]: The StandardCharsets.UTF_8.equals(Charset.forName(str)) would include "utf-8", "UTF-8", "utf8" and "UTF8", but last two one are not supported in Spark4.0. I'd suggest we keep current comparison.

File: spark/src/main/spark-4.x/org/apache/comet/shims/ShimCometExprs.scala (hunk 4 above)
Comment (YutaLin) [reply spark/src/main/spark-4.x/org/apache/comet/shims/ShimCometExprs.scala]: I've rename it `CometExprShimCommon`.

File: spark/src/main/spark-4.1/org/apache/comet/shims/CometExprShim.scala (hunk 6 above)
Comment (YutaLin) [reply spark/src/main/spark-4.1/org/apache/comet/shims/CometExprShim.scala:68]: Yes, thanks for reminder!

**Comment by YutaLin** [comment]
Hi @coderfender 
Thanks for the review, i've made the change, please help me review again!

**Comment by YutaLin** [comment]
Hi @coderfender
Could you help me check this again? Thanks!

**Comment by coderfender** [comment]
Sorry for the delay. Let me look into it shortly @YutaLin

**Comment by andygrove** [comment]
Thanks @YutaLin. This looks good. Could you fix the conflicts?

**Comment by YutaLin** [comment]
Hi @andygrove, 
thanks for the review! I've fix the conflict!

**Comment by coderfender** [comment]
Seems like there are still some more conflicts @YutaLin  .

**Comment by YutaLin** [comment]
Hi @coderfender
Thanks for pointing that out. The previous conflict had already been resolved and reviewed. Since the PR wasn't merged immediately afterward, additional changes on the main branch introduced a new conflict. I've now merged the latest main branch and resolved that conflict as well.

Could you please take another look when you have a chance?

**Comment by coderfender** [comment]
Sure. Yeah have been working on some doc changes causing merge conflicts. I will review it shortly

**Review by andygrove — COMMENTED** [review]
Nice approach lowering `encode(col, 'utf-8')` to a binary cast instead of writing a native implementation. Most of the earlier feedback looks addressed (null charset guard, the utf8 alias discussion, the trait rename, the 4.x consolidation), so I only have one substantive point plus a couple of minor ones.

**Invalid UTF-8 input.** The equivalence "`encode(str, 'utf-8')` equals `cast(str AS binary)`" only holds when the input string is valid UTF-8. Spark's `Encode.encode` guards the raw-bytes fast path with `input.isValid`:

```scala
if ("UTF-8".equalsIgnoreCase(toCharset) && input.isValid) return input.getBytes
```

When a `UTF8String` holds malformed UTF-8, Spark does not return the raw bytes. It routes through a `CharsetEncoder` and, with the default non-legacy error action, either substitutes or throws `malformedCharacterCoding`. Spark 3.x reaches a similar outcome via `input.toString.getBytes(charset)`, which turns invalid sequences into the replacement character first.

The cast approach here always returns the raw underlying bytes, so for a valid-UTF-8 column the two match perfectly, but for an invalid-UTF-8 string Comet would silently produce a different result than Spark's `encode`, with no fallback since the charset is still `utf-8`. Invalid UTF-8 in a `StringType` column is uncommon but reachable, for example via `cast(binary_col as string)` over non-UTF-8 bytes. Since `expressions.md` marks encode as fully compatible, could we either note this as a known limitation or add a test that feeds an invalid-UTF-8 string through both engines to confirm the divergence? At minimum a comment next to the "byte-equivalent to cast" note pointing out the `isValid` caveat would help the next reader.

**StaticInvoke match robustness (minor).** The 4.x match compares `arguments.size == 4` and the full `inputTypes` sequence including `StringTypeWithCollation(supportsTrimCollation = true)`. If a future 4.x minor version tweaks that shape, the match silently fails and encode falls back. That is safe but easy to miss. A looser match keyed on `staticObject == classOf[Encode]` and `functionName == "encode"` with positional argument extraction might be more robust across versions, if that is acceptable.

**Test file (minor).** `encode.sql` is missing a trailing newline, which some lint or RAT steps dislike. An invalid-UTF-8 input case would also directly probe the compatibility question above.

Overall this looks close. The invalid-UTF-8 behavior is really the only item I would want to settle before merge.

**Comment by comphead** [comment]
is this PR related to https://github.com/apache/datafusion/pull/21331 ?

**Comment by YutaLin** [comment]
Hi @andygrove 
Thanks for reviewing this. I kept the cast lowering for valid UTF-8, documented the malformed UTF-8 limitation next to the implementation and in the compatibility guide, and captured the reproduction under #4764. I also loosened the 4.x StaticInvoke match and fixed the trailing newline.

**Comment by andygrove** [comment]
> **Note on this review:** this was generated by an LLM (Claude Code) at my request while I worked through a review backlog. I have not verified the individual findings myself. Please treat everything below as suggestions to evaluate rather than as authoritative review feedback, and push back on anything that is wrong or already handled.

Lowering `encode(str, 'utf-8')` to a `CAST(string AS binary)` is a neat way to get this for free, and the `CometExprShimCommon` trait shared across the 4.x shims is the right structure for the `StaticInvoke` rewrite. The SQL fixture covers empty strings, NULL, multibyte, and the mixed-case charset literal, which is good.

Three things.

**The malformed-UTF-8 divergence is not reported to users**

Spark replaces malformed bytes during `encode`, while the cast lowering preserves them. That is a wrong answer, not a fallback, and right now the only record of it is an `ignore(...)` line in `encode.sql` plus a code comment.

`encode` should report `Incompatible(Some(...))` for this, or at minimum the divergence needs to appear on the string compatibility page so a user can find it. As it stands, `expressions.md` will show `encode` as supported with no caveat and a user with dirty string data gets silently different bytes. That is the exact case Comet's `Incompatible` mechanism exists for.

If you would rather not gate it, that is a discussion worth having explicitly, but it should not be an implicit consequence of where the code happens to live.

**Charset alias matching**

```scala
str.toString.toLowerCase(Locale.ROOT) == "utf-8"
```

Java's `Charset.forName` accepts `UTF8`, `utf8`, and `unicode-1-1-utf-8` as aliases for UTF-8, and Spark accepts whatever `Charset.forName` accepts. So `encode(s, 'UTF8')` falls back here even though it is exactly the case this PR handles.

Would `Try(Charset.forName(name).name() == "UTF-8").getOrElse(false)` be better? It handles aliases, and an invalid charset name naturally falls back to Spark, which then raises the proper `INVALID_PARAMETER_VALUE`.

**Duplication between the 3.4 and 3.5 shims**

The additions to `spark-3.4/CometExprShim.scala` and `spark-3.5/CometExprShim.scala` are identical. The 4.x side already shares through `CometExprShimCommon`. Is there a `spark-3.x` shared location that could host the same dispatch for 3.4 and 3.5? Two copies of a three-line match is minor now, but it is the pattern that makes per-version gaps show up later, and CI only lints a subset of the version matrix.
